java/src/main/native/src/row_conversion.cu

```diff
@@ -2168,41 +2253,106 @@ std::unique_ptr<table> convert_from_rows(lists_column_view const &input,
             {gpu_tile_infos.data() + tile_offset, gpu_tile_infos.size() - tile_offset},
             gpu_batch_row_boundaries, start_col, end_col, tile_height, num_rows, stream);
       });
 
-  dim3 blocks(util::div_rounding_up_unsafe(gpu_tile_infos.size(), NUM_TILES_PER_KERNEL_FROM_ROWS));
-  dim3 threads(NUM_THREADS);
+  dim3 const blocks(
+      util::div_rounding_up_unsafe(gpu_tile_infos.size(), NUM_TILES_PER_KERNEL_FROM_ROWS));
+  dim3 const threads(NUM_THREADS);
 
+  // validity needs to be calculated based on the actual number of final table columns
   auto validity_tile_infos =
-      detail::build_validity_tile_infos(num_columns, num_rows, shmem_limit_per_tile, row_batches);
+      detail::build_validity_tile_infos(schema.size(), num_rows, shmem_limit_per_tile, row_batches);
 
   auto dev_validity_tile_infos = make_device_uvector_async(validity_tile_infos, stream);
 
-  dim3 validity_blocks(
+  dim3 const validity_blocks(
       util::div_rounding_up_unsafe(validity_tile_infos.size(), NUM_VALIDITY_TILES_PER_KERNEL));
 
-  dim3 validity_threads(
+  dim3 const validity_threads(
       std::min(validity_tile_infos.size() * NUM_VALIDITY_THREADS_PER_TILE, 128lu));
 
-  detail::fixed_width_row_offset_functor offset_functor(fixed_width_size_per_row);
+  if (dev_string_row_offsets.size() == 0) {
+    detail::fixed_width_row_offset_functor offset_functor(size_per_row);
 
-  detail::copy_from_rows<<<blocks, threads, total_shmem_in_bytes, stream.value()>>>(
-      num_rows, num_columns, shmem_limit_per_tile, offset_functor, gpu_batch_row_boundaries.data(),
-      dev_output_data.data(), dev_col_sizes.data(), dev_col_starts.data(), gpu_tile_infos,
-      child.data<int8_t>());
+    detail::copy_from_rows<<<blocks, threads, total_shmem_in_bytes, stream.value()>>>(
+        num_rows, num_columns, shmem_limit_per_tile, offset_functor,
+        gpu_batch_row_boundaries.data(), dev_output_data.data(), dev_col_sizes.data(),
+        dev_col_starts.data(), gpu_tile_infos, child.data<int8_t>());
 
-  detail::copy_validity_from_rows<<<validity_blocks, validity_threads, total_shmem_in_bytes,
-                                    stream.value()>>>(
-      num_rows, num_columns, shmem_limit_per_tile, offset_functor, gpu_batch_row_boundaries.data(),
-      dev_output_nm.data(), column_info.fixed_width_column_starts.back(), dev_validity_tile_infos,
-      child.data<int8_t>());
+    detail::copy_validity_from_rows<<<validity_blocks, validity_threads, total_shmem_in_bytes,
+                                      stream.value()>>>(
+        num_rows, num_columns, shmem_limit_per_tile, offset_functor,
+        gpu_batch_row_boundaries.data(), dev_output_nm.data(), column_info.column_starts.back(),
+        dev_validity_tile_infos, child.data<int8_t>());
+
+  } else {
+    detail::string_row_offset_functor offset_functor(device_span<size_type const>{input.offsets()});
+    detail::copy_from_rows<<<blocks, threads, total_shmem_in_bytes, stream.value()>>>(
```

Comment:
we have to be careful, because we're using the `copy_from_rows` kernel to copy the offsets and lengths for the string kernel, but I believe we could do this on `copy_to_rows`. The validity kernels can run concurrently for sure. I'll add an issue for it.

Reply:
Added issue https://github.com/NVIDIA/spark-rapids-jni/issues/274